src/LanguageServer/Protocol/LspServices/LspServices.cs

```diff
@@ -40,36 +41,48 @@ public LspServices(
     {
         var serviceMap = new Dictionary<string, Lazy<ILspService, LspServiceMetadataView>>();
 
-        // Convert MEF exported service factories to the lazy LSP services that they create.
-        foreach (var lazyServiceFactory in mefLspServiceFactories)
-        {
-            var metadata = lazyServiceFactory.Metadata;
+        // Add services from factories exported for this server kind.
+        foreach (var lazyServiceFactory in mefLspServiceFactories.Where(f => f.Metadata.ServerKind == serverKind))
+            AddSpecificService(new(() => lazyServiceFactory.Value.CreateILspService(this, serverKind), lazyServiceFactory.Metadata));
 
-            // Make sure that we only include services exported for the specified server kind (or NotSpecified).
-            if (metadata.ServerKind == serverKind ||
-                metadata.ServerKind == WellKnownLspServerKinds.Any)
-            {
-                serviceMap.Add(
-                    metadata.TypeRef.TypeName,
-                    new(() => lazyServiceFactory.Value.CreateILspService(this, serverKind), metadata));
-            }
-        }
+        // Add services exported for this server kind.
+        foreach (var lazyService in mefLspServices.Where(s => s.Metadata.ServerKind == serverKind))
+            AddSpecificService(lazyService);
 
-        foreach (var lazyService in mefLspServices)
-        {
-            var metadata = lazyService.Metadata;
+        // Add services from factories exported for any (if there is not already an existing service for the specific server kind).
+        foreach (var lazyServiceFactory in mefLspServiceFactories.Where(f => f.Metadata.ServerKind == WellKnownLspServerKinds.Any))
+            TryAddAnyService(new(() => lazyServiceFactory.Value.CreateILspService(this, serverKind), lazyServiceFactory.Metadata));
 
-            // Make sure that we only include services exported for the specified server kind (or NotSpecified).
-            if (metadata.ServerKind == serverKind ||
-                metadata.ServerKind == WellKnownLspServerKinds.Any)
-            {
-                serviceMap.Add(metadata.TypeRef.TypeName, lazyService);
-            }
-        }
+        // Add services exported for any (if there is not already an existing service for the specific server kind).
+        foreach (var lazyService in mefLspServices.Where(s => s.Metadata.ServerKind == WellKnownLspServerKinds.Any))
+            TryAddAnyService(lazyService);
 
         _lazyMefLspServices = serviceMap.ToFrozenDictionary();
 
         _baseServices = baseServices;
+
+        void AddSpecificService(Lazy<ILspService, LspServiceMetadataView> serviceGetter)
```

Comment:
>void [](http://example.com/codeflow?start=8&length=4)

personal preference: I find local fns easier to follow if they are static

Reply:
left as-is - in this specific case I prefer the non-static version so it can easily act on serviceMap